spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/autoconfig/GcpCloudSqlAutoConfiguration.java

```diff
@@ -0,0 +1,123 @@
+import org.springframework.cloud.gcp.sql.CloudSqlJdbcUrlProvider;
+import org.springframework.cloud.gcp.sql.GcpCloudSqlProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.util.StringUtils;
+
+/**
+ * Google Cloud SQL starter.
+ *
+ * <p>It simplifies database configuration in that only an instance and database name are provided,
+ * instead of a JDBC URL. If the instance region isn't specified, this starter attempts to get it
+ * through the Cloud SQL API.
+ *
+ * @author João André Martins
+ */
+@Configuration
+@EnableConfigurationProperties(GcpCloudSqlProperties.class)
+@ConditionalOnClass({GcpProperties.class, SQLAdmin.class})
+@AutoConfigureBefore(DataSourceAutoConfiguration.class)
+@AutoConfigureAfter(GcpContextAutoConfiguration.class)
+public class GcpCloudSqlAutoConfiguration {
+
+	@Autowired
+	private GcpCloudSqlProperties gcpCloudSqlProperties;
+
+	@Autowired
+	private GcpProperties gcpProperties;
+
+	@Bean
+	@ConditionalOnMissingBean
+	protected SQLAdmin sqlAdminService(CredentialsProvider credentialsProvider)
+			throws GeneralSecurityException, IOException {
+		HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+		JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+
+		return new SQLAdmin.Builder(httpTransport, jsonFactory,
+				new HttpCredentialsAdapter(credentialsProvider.getCredentials()))
+				.setApplicationName("Google-SQLAdminSample/0.1")
+				.build();
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(CloudSqlJdbcUrlProvider.class)
+	public CloudSqlJdbcUrlProvider cloudSqlJdbcUrlProvider(SQLAdmin sqlAdmin) throws IOException {
+		// Look for the region if the user didn't specify one.
+		String region = this.gcpCloudSqlProperties.getRegion();
+		if (!StringUtils.hasText(region)) {
+			region = sqlAdmin.instances().get(this.gcpProperties.getProjectId(),
+					this.gcpCloudSqlProperties.getInstanceName()).execute().getRegion();
+		}
+
+		String instanceConnectionName = this.gcpProperties.getProjectId() + ":"
+				+ region + ":" + this.gcpCloudSqlProperties.getInstanceName();
+
+		return () -> String.format("jdbc:mysql://google/%s?cloudSqlInstance=%s&"
+						+ "socketFactory=com.google.cloud.sql.mysql.SocketFactory",
+				this.gcpCloudSqlProperties.getDatabaseName(),
+				instanceConnectionName);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(DataSource.class)
+	public DataSource cloudSqlDataSource(CloudSqlJdbcUrlProvider cloudSqlJdbcUrlProvider)
+			throws GeneralSecurityException, IOException, ClassNotFoundException {
+		// TODO(joaomartins): Set the credential factory to be used by SocketFactory when
+		// https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/issues/41 is
+		// resolved. For now, it uses application default creds.
+
+		// Load MySQL driver to register it, so the user doesn't have to.
+		Class.forName("com.mysql.jdbc.Driver");
```

Comment:
Yeah, you're right actually. The magic we're really adding is the URL builder, the user should have more freedom to choose their `DataSource`.